### DISCUSSION TITLE
Possible widget enhancements

### DISCUSSION THREAD

**Discussion by unalignedcoder** [TechbeeAT/jtxBoard#285] [General]
You guys have probably planned similar enhancements and more, so this may be redundant. Anyway - 

It would be great if the widget:

1) Instead of having a title that says simply "tasks", it showed first the name of the Collection the items belong to. In case the user has different widgets for different collections;

2) The ability to choose colors;

3) The ability to choose the corner radius.

**Comment by patrickunterwegs** [comment]
Hey @unalignedcoder , 

the Widget Header is configurable since 2.2.1, see #232 
I have been thinking about color options, but it is not a killer-feature with high priority. Anyway, I have made a feature entry for that: #286

**Reply by patrickunterwegs** [reply]
I have put some more thoughts into the color choosing, it is not as trivial as it sounds:

1) There are two background colors defined, one for the widget background, one for the entry background. When a user can choose the colors, dark and light mode need to be considered. So to make it properly, the user should be able to choose colors for both modes, so 4 in total. 
2) If you let the user choose any color, the text color might not fit. So there's the option to let the user choose (again 2 x2 colors for the text on the background and the text on the entry), which already makes 8 colors in total to choose. This can end up in a mess. The alternative would be to set the text color automatically. In Jetpack Compose there is an option to calculate the best matching color for a text based on a background color. This function is not available yet in the widget library and might come at a later stage. 

While 1) would still be acceptable to do (make 4 color selectors), for 2) I would only consider the automatic calculation of the text color. But this depends on when Google also provides the color calculation for the widget library.

**Comment by unalignedcoder** [comment]
I see. It's not an easy problem to fix, given the multitude of user choices.

Perhaps just allowing the user **control over the title text color** would be enough.

Personally, I don't appreciate the bluish dark color the title text takes on a Light Theme.

I would appreciate being able to permanently (regardless the System Theme) set the text title color to **white with a bit of shadow**, so that it could be read on brighter backgrounds as well. 

Sort of like the system clock widget does (at least on Android 10 where I am).

Also allowing the user to tweak the size would be good.

Bottom line: **In absence of more advanced solutions, imitating the system widgets may be the way to go.**

**Reply by patrickunterwegs** [reply]
> Bottom line: In absence of more advanced solutions, imitating the system widgets may be the way to go.

This is what the app actually does. I mean, what are system widgets anyway? Samsung is also offering kind of system widgets that are not very consistent with other widgets.
The colors are defined based on the Material3 colors, which follows the latest definitions for system colors (which is probably more up-to-date than many other apps). It also supports dynamic colors, so if you have Android 12+ you can just change the colors in your system settings. The text color is taken from the calculated colors of the base Material3 theme. (Note: The text color is calculated implicitly when the color or opacity cannot be changed by the user)

I could change the title color to pure white for dark mode and pure black for dark mode, but then the app wouldn't follow the Material3 Design colors anymore...

**Reply by unalignedcoder** [reply]
It is a very minor problem! And I don't know about the Material3 standards; but I am not sure why it should generate a title in purple-blue, while the rest of the widget stays safely in black and white...

**Reply by patrickunterwegs** [reply]
I will check it. Maybe I can use the text color there as well.
But for a quick explanation: 
Material3 defines different colors for example Primary, Surface, PrimaryContainer,... that are used on different elements.
Most of these colors have a Text color defined (or calculated), e.g. onPrimary, onSurface,...

When the background used PrimaryContainer, it would be expected to use the onPrimaryContainer color which just looks the way you see it. You just can't see other color schemes because Android does not support dynamic colors. 
Anyway, I'll check how it looks with the same color as the items text in the header!

**Reply by patrickunterwegs** [reply]
I have implemented a color picker now. There are still some downsides, so I added a  disclaimer, e.g. day/night changes are not reflected immediately when changing the opacity, when setting a color, the day/night change will be completely ignored.
I didn't do anything about the corner radius, I believe that goes really a step too far with configurations. Currently there is a corner radius defined, but older versions of android just don't support it, so older android versions might see sharp edges (a dedicated config wouldn't change this anyway!)
